Reuse recommendation

nr23730 opened the pull request:
<img width="490" alt="Screenshot 2022-05-05 at 13 26 24" src="https://user-images.githubusercontent.com/20756025/166913921-2df959d6-1772-40fe-a823-96c2fc8ce68d.png">
<img width="605" alt="Screenshot 2022-05-05 at 13 58 01" src="https://user-images.githubusercontent.com/20756025/166918499-60585edd-acc2-42ef-b3cc-112092b506d1.png">

Unberath commented on src/pages/patientView/therapyRecommendation/form/TherapyRecommendationFormOtherMtb.tsx:
```diff
@@ -0,0 +1,169 @@
+                        <Modal.Title>{this.props.title}</Modal.Title>
+                    </Modal.Header>
+                    <Modal.Body>
+                        Did not find any mathing recommendations from other MTB
```
Spotted a typo: mathing -> matching

Unberath commented:
Great addition!
I'm not sure how we can handle nearly identical prior recommendations. If two patients both got the same recommendation but with different comments/clinical data, there's no way to distinguish them in this form, right?
Ideally we would want to group these and combine them to a more general recommendation but that's for another project.
Maybe add the date of the MTB or patient ID? Tooltip could also work.

Unberath commented:
@arsenij-ust should also take a look at this

arsenij-ust commented:
@nr23730 Realy nice work! Philipp's point is indeed important. A tooltip is one possibility. Another idea is that when saving the MTB recommendation that the user wants to reuse, the user can set a short comment or name that is unique and is displayed in the drop-down menu. Are all MTB recommendations automatically saved or can the user decide whether a recommendation is included in the "reuse list"? The recommendations in the "reuse list" serve as templates - I can well imagine that the clinicians would like to have the possibility to edit or delete the recommendation templates afterwards. Perhaps a separate tab for the management of user-defined recommendation templates would be useful for this. So this intertwines strongly with Philipp's approach towards own OnkoKB/Muriel.

arsenij-ust commented:
A small note on the title of the drop-down menu. Maybe it would be better to write - "Add therapy recommendation from templates" instead of "Add therapy recommendation from other MTB".

Unberath commented:
> Another idea is that when saving the MTB recommendation that the user wants to reuse, the user can set a short comment or name that is unique and is displayed in the drop-down menu. Are all MTB recommendations automatically saved or can the user decide whether a recommendation is included in the "reuse list"? The recommendations in the "reuse list" serve as templates - I can well imagine that the clinicians would like to have the possibility to edit or delete the recommendation templates afterwards. Perhaps a separate tab for the management of user-defined recommendation templates would be useful for this. So this intertwines strongly with Philipp's approach towards own OnkoKB/Muriel.

These are all important remarks, but I'm afraid that the whole issue of templating is too complex for this PR. 
Personally, I would leave this PR for reusing already documented recommendations and keep discussion open for the next project (templating/adapted OncoKB).

nr23730 commented:
> A small note on the title of the drop-down menu. Maybe it would be better to write - "Add therapy recommendation from templates" instead of "Add therapy recommendation from other MTB".

This is addressed with bc3d1155e7ba0ebb60691e16b529885c4bfbc952.

nr23730 commented:
General notes on this:

The clinical attributes are not transmitted from FhirSpark, because they are specific for a patient.
Right now FhirSpark is implemented to search in all therapy recommendations. As @Unberath mentioned, building a discrete highly curated knowledge base was not the intention of this PR. 

My plan would now be to add the tooltip to provide additional information on the recommendation.

As @Unberath mentioned: A highly curated expert system is probably out of scope for this project phase.

nr23730 commented:
@Unberath I added a tooltip that shows the author. For now I don't show any patient id or clinical data of the other recommendation as those won't be sent by FhirSpark to provide a little more privacy. 
If you want to have this included, I'll have to add this in Fhirspark first.

Unberath commented:
@nr23730 I got it to work but I'm getting an error when the switch of the new modal to the edit modal is happening:
<img width="703" alt="image" src="https://user-images.githubusercontent.com/33686055/176439916-4c71e2bb-14c4-4502-b74f-6e5b2b6a6e77.png">

I think I found the issue, but I'm not sure and I think you're having the better overview anyways:
Is the alteration/mtb endpoint already adapted for clinical trials as recommendations? 

But for I know (as I'm testing with the remote FHIRspark setup) this could also be just a version mismatch?

nr23730 commented:
Good catch! This was introduced by the clincial trial clipboard. I was able to fix this using Fhirspark. Now everything should work smooth. Just pushed a new docker image for FhirSpark.

Unberath commented:
I think this feature is a bit more complex than we originally thought. The described error is fixed, but unfortunately there are some more things going wrong on my side 🥲:
- Deleted recommendations still show up in the selection.
- How complex would it be to extend the tooltip with the date of documentation? Even with my 2-3 sample datasets, when removing and creating the same recommendation I didn't knew which was which (I guess for the real world we don't need a timestamp, just the date)
- Recommended trials get lost if the recommendation is added from template
- There is some issue with evidence levels which I'm not so sure where it's coming from (also when the switch of the new modal to the edit modal is happening):
<img width="703" alt="image" src="https://user-images.githubusercontent.com/33686055/176461047-bf848e01-6256-456f-9955-8a28dc8a8e4c.png">


I used these sample data for testing: http://131.188.43.108:8080/patient/mtb?studyId=testpatient&caseId=Testpatient

nr23730 commented:
Okay, let me sort this a little bit:

> Deleted recommendations still show up in the selection.

This is an issue with FhirSpark where it seems that only the MTB is deleted / the Therapy recommendation is removed from the MTB session. Update: Also fixed this in FhirSpark. New Docker image uploaded.

> How complex would it be to extend the tooltip with the date of documentation? Even with my 2-3 sample datasets, when removing and creating the same recommendation I didn't knew which was which (I guess for the real world we don't need a timestamp, just the date)

Well it's not directly attached to the therapy recommendation in our current model but I still see a way to get this fixed (also has to be done in FhirSpark). Update: Done!

> Recommended trials get lost if the recommendation is added from template

Also an issue with FhirSpark that does not query them yet. Update: Done!

>There is some issue with evidence levels which I'm not so sure where it's coming from (also when the switch of the new modal to the edit modal is happening):

Yeah, here the frontend code expects something that is not there. Found the issue and fixed it.

Unberath commented:
Thank you so much for all the fixing!

 
> > Deleted recommendations still show up in the selection.
> 
> This is an issue with FhirSpark where it seems that only the MTB is deleted / the Therapy recommendation is removed from the MTB session. Update: Also fixed this in FhirSpark. New Docker image uploaded.

Did you push multiple image versions? I did pull a new version, but I still see deleted recommendations (even ones I created and deleted after the update). Or do I need to do more than just pulling FhirSpark?

And one additional questions: Your selection only supports (by design?) mutations and not CNVs, right? 

Everything else works great!

nr23730 commented:
> Did you push multiple image versions? I did pull a new version, but I still see deleted recommendations (even ones I created and deleted after the update). Or do I need to do more than just pulling FhirSpark?

Well, this only works if you now delete a therapy recommendation. Old (halfway deleted) therapy recommendations will not be affected by this change. This image is the *usual* one so `ghcr.io/buschlab/fhirspark:latest`.

> And one additional questions: Your selection only supports (by design?) mutations and not CNVs, right?

Yes, by now this should only support variants and not CNVs.

Unberath commented:
> > Did you push multiple image versions? I did pull a new version, but I still see deleted recommendations (even ones I created and deleted after the update). Or do I need to do more than just pulling FhirSpark?
> 
> Well, this only works if you now delete a therapy recommendation. Old (halfway deleted) therapy recommendations will not be affected by this change. This image is the _usual_ one so `ghcr.io/buschlab/fhirspark:latest`.

I really don't know if there's something wrong on my side but I still see the therapy recommendations I created and deleted after the FhriSpark update. 

Another thing I just found is that the search doesn't work at all (adapted filterOptions from TherapyRecommendationFormOncoKb should be an easy fix). I don't know if you want to postpone some refactoring for after the release, I'm really sorry this keeps getting longer.

nr23730 commented:
Okay, added now search using filterOptions and fixed deleting therapy recommendations. 
Issue was that cBioPortal did not preserve the correct order (first save mtb so that reference is removed, than delete therapy recommendation).

Unberath reviewed (approved):
lgtm, thank you so much!
